### DISCUSSION TITLE
Adding a folder from one repo to another

### DISCUSSION THREAD

**Discussion by CoreMediaVini** [community/community#23346] [New to GitHub]
<p>I have been playing around to create a <code>_subtree_</code>.</p>
<pre><code>RepoFolderSrc RepoFolderTarget RepoAnother
        | | |
      develop sfcc------------------sfcc18
        | |
        | |
        |-someFolder1 |
        |-someFolder2----------featureBranch
</code></pre>
<p>I want to copy the <code>_someFolder1_</code> and <code>_someFolder2_</code> from <code>_RepoFolderSrc_</code> to <code>_RepoFolderTarget_</code>. I want to retain all the history for the folder <code>_someFolder1_</code> and <code>_someFolder2_</code> when it is copied over to the <code>_featureBranch_</code>.</p>
<p>A few of the solutions I have seen requires deleting the <code>_origin_</code> from <code>_RepoFolderSrc_</code>. I should not make any changes to the <code>_RepoFolderSrc_</code> and I should not create any additional branch or commits on the folder.</p>
<p>Is there a way by which I can copy over the folder by only playing around with the <code>_RepoFoldertarget_</code>. I do have push and pull permissions for all repositories.</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>If you’re referring to <a href="http://gbayer.com/development/moving-files-from-one-git-repository-to-another-preserving-history/" rel="nofollow noopener">the solution described in this blog post</a>, as one of the ones you’ve seen but disliked how it worked because it made changes to the source repository, then I’m afraid to say that it is the canonical solution for getting done what you’re describing. I also think you’re misunderstanding when you say that the solution requires "deleting the <code>origin</code>" or “create additional branch or commits”. Let me see if I can rephrase the solution more clearly.</p>
<p>The goal here is to copy from a repo named <code>source</code> to a repo named <code>destination</code> only the contents of folder <code>foo</code>, including all history that touches folder <code>foo</code>. In order to do that, because Git doesn’t track folders <em>per se</em> but tracks commits, we’ll have to filter the commit history of the <code>source</code> repo to only contain the commits that touch folder <code>foo</code>. The way we’ll do that is via the <a href="https://git-scm.com/docs/git-filter-branch" rel="nofollow noopener"><code>git filter-branch</code> command</a> using the <code>--subdirectory-filter</code> filter. First, we’ll need to create a local copy of the <code>source</code> repo  <strong>separate from</strong> any local copy you already have. You can do this by doing the following:</p>
<pre><code>mkdir backup
cd backup
git clone https://github.com/user/source.git
cd source
</code></pre>
<p>Now, we are in the directory <code>backup/source</code> which is a completely new local copy of the <code>source</code> repo separate from any local copy you already have. Additionally, we want to <strong>protect</strong> <a href="https://github.com/user/source" rel="nofollow noopener">https://github.com/user/source</a> from being accidentally changed by our process here. We do that by deleting the <code>origin</code> remote,  <strong>but only in this new local copy.</strong></p>
<pre><code>git remote rm origin
</code></pre>
<p>Next, we want to filter the commit history of the <code>source</code> repo to only contain the commits that touch folder <code>foo</code>. We do that by issuing the <code>filter-branch</code> command:</p>
<pre><code>git filter-branch --subdirectory-filter foo -- --all
</code></pre>
<p>But, as you’ll notice in the documentation for the <code>git filter-branch --subdirectory-filter</code> command, it states:</p>
<blockquote>
<p>The result will contain that directory (and only that) as its project root.</p>
</blockquote>
<p> This means that everything that  <strong>was</strong> in <code>foo</code> is now in the root of the repo. So, in order to put it all back in folder <code>foo</code>, we have to add one commit to this  <strong>local only temporary working repository</strong> that we’re going to throw away soon:</p>
<pre><code>mkdir foo
mv * foo
git add .
git commit
</code></pre>
<p>Now we have a local repository that contains only the files and history that we want. We’re halfway there! :tada:And remember, this is just a temporary place to store some work. It is completely separated from your normal local <code>source</code> repository and from <a href="https://github.com/user/source" rel="nofollow noopener">https://github.com/user/source</a>, so any changes we make here are not permanent. It’s like making a temporary backup of a file so that you can revert back to the original if you screw something up.</p>
<p>Now, in order to merge this new history that we’ve created into the destination repository, let’s assume that the destination repository is already cloned locally in <code>projects/destination</code> from <a href="https://github.com/user/destination" rel="nofollow noopener">https://github.com/user/destination</a>. So we need to:</p>
<pre><code>cd ../../projects/destination
</code></pre>
<p>in order to move from <code>backup/source</code> to <code>projects/destination</code>. Then we issue the following commands:</p>
<pre><code>git remote add modified-source ../../backup/source
git pull modified-source master --allow-unrelated-histories
</code></pre>
<p>This pulls all of the commit history of our <strong>temporary</strong>  <code>source</code> repo into the <code>master</code> branch of our local copy of the <code>destination</code> repo. Then all that is needed is to clean things up:</p>
<pre><code>git remote rm modified-source
</code></pre>
<p>This removes the link between our local <code>destination</code> repo and the temporary <code>source</code> repo. And when you’re sure everything worked correctly, you can delete the modified local <code>source</code> repo whenever you feel like.</p>
<p>I hope that helps explain things a bit better! Let me know if you have any questions.</p>

**Reply by gonzales-jorge-siul** [reply]
So helpful, thanks.  

I just noted that the filter-branch has a warning, it recommends to use git-filter-repo. So, instead I used: `git filter-repo --subdirectory-filter foo --force`. I used `--force` because there is check for the origin and we removed.

By the way you need to install git-filter-repo to be able to use it.

**Comment by qingbol** [comment]
<p>very detailed and targeted answer, this is really useful</p>

**Comment by vinischeidegger** [comment]
<p>Really good answer. From my understanding, this will copy the most recent commit and the history for the specific subfolder. Is there a way to keep the source remote, and thus the hability to sync only that folder?</p>
<p>Thank you</p>

**Comment by lee-dohm** [comment]
<p>You can’t keep the source remote and sync only that folder, no. But, you could remove the source folder from the original repository and add the new repository back to the original one as a <a href="https://git-scm.com/book/en/v2/Git-Tools-Submodules" rel="nofollow noopener">submodule</a>. I believe this would provide essentially the functionality you’re looking for.</p>
<p>I hope that helps!</p>

**Comment by agwatic** [comment]
<p>Hello <a class="mention" href="/lee-dohm">@lee-dohm</a> ,</p>
<p>If I want to wrap your answer in a batch file, is there any way to revert the old repo to its original state before running this batch file?</p>

**Comment by lee-dohm** [comment]
<p>I’m not sure I understand the question <a class="mention" href="/agwatic">@agwatic</a>. The whole point of my answer is for it to not change the original repo in any way, so I’m not sure what there would be to revert?</p>

**Comment by utkucan** [comment]
<p>Hi <a class="mention" href="/lee-dohm">@lee-dohm</a> </p>
<pre><code>git filter-branch --subdirectory-filter foo -- --all
</code></pre>
<p>We had a similar issue, and used your solution but ended up losing part of history.</p>
<p>The folder we wanted to move to another repo was carried from different directory one year ago. In current repository we are able to see whole history but when I use filter-branch and extract the folder, the history before directory change was carried out is gone.</p>
<p>Do you have any ideas how we can keep whole history for this folder?</p>
<p>Thanks.</p>

**Comment by lee-dohm** [comment]
<p>Unfortunately, I do not have any ideas <a class="mention" href="/utkucan">@utkucan</a>. I assume that the history before the directory change was carried out is gone because you’re using the <code>--subdirectory-filter</code> option and specifying the current directory. This means that the old directory is specifically filtered out. From <a href="https://git-scm.com/docs/git-filter-branch#git-filter-branch---subdirectory-filterltdirectorygt" rel="nofollow noopener">the documentation</a> it doesn’t appear that specifying multiple directories is an option.</p>
<p>You may want to experiment with other filter options to see if you can find something that would work the way you want.</p>

**Comment by agwatic** [comment]
<p>After making your tutorial I didn’t manage to get back to the old state of the repo. How can I get back to the old state before extracting the directory from the repo?</p>

**Comment by lee-dohm** [comment]
<p>As I said before, the process I described <em>does not change</em> the original repository. If your original repository was changed, can you tell me how exactly it was changed?</p>

**Comment by BryanLa** [comment]
<p>Thanks <a class="mention" href="/lee-dohm">@lee-dohm</a> .</p>
<p>I’m using this exact method. But doing the following step in the “source” repo, causes all commit history to be lost/unviewable when I push my “destination” repo branch back up to GitHub:</p>
<pre><code>mkdir foo
mv * foo
git add .
git commit
</code></pre>
<p>If I omit that step, I can view all of the commit history in GitHub, in tact.</p>
<p>Is there any way possible to reintroduce the original “foo” directory structure in the “destination” repo, and preserve commit history in GitHub? I know there are a lot of posts about this, indicating that <span> git log </span> will show the original history, and I’ve confirmed that as well. But is there a way to cause GitHub to show it?</p>
<p>Thanks</p>

**Comment by mukkuonline** [comment]
<p>Hi.This is very nice demo but there is one problem which I am struggling and didn’t find when I started google since last week. The file which I need to copy from source repo to target repo is actually copied in the root folder. I am unable to find which option i must use if i want to copy to specific folder at destination repo. eg. I have a file ‘myfile.txt’ in /sourcerepo/srcfolder/srcsubfolder and want to copy into /targetrepo/targetfolder/targetsubfolder. I tried so many options and asked so mamy friends, everyone treid but file copied only in /targetrepo not in /targetrepo/targetfolder/targetsubfolder. Please help me how to copy in targetsubfolder or where should i give path for copy file in target folder. I appreciate all for your help. Thanks.</p>

**Comment by theclapp** [comment]
<p>Could you not do the same process again, this time with the older directory?</p>

**Comment by lukasz-r** [comment]
<p>A very nice tutorial.</p>
<p>Things can be simplified a bit by the use of the <strong>git subtree</strong> command. You can extract a history related to a subdirectory into a branch:</p>
<pre><code>git subtree split -P subdir --annotate="(split)" -b split
</code></pre>
<p>This way you end up with the <strong>split </strong> branch holding a history related to the <strong>subdir</strong> directory, with each commit prefixed with the <strong>(split)</strong> string. Then you can use standard branch operation tools to push/merge the branch, etc. A good thing is that you can safely do it in the main repo without fear of destroying anything.</p>

**Comment by johwer** [comment]
<p>Great answer and work!</p>
<p>My complained about this line in terminal but it with a look at files and folder it had done it is job.</p>
<pre><code>mv * foo
</code></pre>
<p>mv: rename foo to foo/foo: Invalid argument</p>

**Comment by haznix** [comment]
<p>I had to do this recently, and I ended up using git-filter-repo. I was moving a subfolder from my current repo in to a branch of another repo. I used the following commands:</p>
<p><strong>Note: This will bring out the history to your repo root and completely re-write it, destroying the old history in the process.</strong></p>
<p><code>git filter-repo --subdirectory-filter [DIRNAME] --force</code></p>
<p>then in the new repo I created a new orphan branch</p>
<pre><code class="lang-auto">git checkout --orphan [BRANCH_NAME]
git reset --hard
</code></pre>
<p><strong>Run the following in the branch of the new repo, assuming you have them both cloned out locally</strong></p>
<pre><code class="lang-auto">git pull [local-path-to-filtered-repo] [BRANCH_NAME] 
git push -o skip-validation [origin/upstream] [BRANCH_NAME]
</code></pre>
<p>Hope this helps.</p>

**Comment by prasanthsarath** [comment]
<p>I think this approach by default moves directory to the target repo’s ROOT level. But how can we move the source folder from repo 1 to a specific target location in repo 2 (not to the ROOT)?</p>

**Comment by prasanthsarath** [comment]
<p>Did you find a solution to your use case? I have the same issue.</p>
